### ISSUE TITLE
Orthonormal bases and vectorized representations

### ISSUE THREAD

**Issue by sethaxen** [JuliaNLSolvers/Manifolds.jl#66]
We have a number of important uses for ONBs. In some cases, given a tangent vector, we want to get a column vector representation with length the same as the dimension of the manifold and don't necessarily care about explicitly working out the bases (an example is the `vee`/`hat` functions for `Rotations`, which explicitly get a vector with respect to the basis or project a vector onto the basis to get a skew symmetric matrix; the basis isn't orthonormal but can be made so by dividing by `√2`)

```
julia> using Manifolds, LinearAlgebra

julia> R = Manifolds.Rotations(3);

julia> x = Matrix{Float64}(I, 3, 3);

julia> v = hat(R, x, [1., 2., 3.])
3×3 StaticArrays.MArray{Tuple{3,3},Float64,2,9} with indices SOneTo(3)×SOneTo(3):
  0.0  -3.0   2.0
  3.0   0.0  -1.0
 -2.0   1.0   0.0

julia> vee(R, x, v)
3-element StaticArrays.MArray{Tuple{3},Float64,1,3} with indices SOneTo(3):
 1.0
 2.0
 3.0

julia> inner(R, x, hat(R, x, [1, 0, 0]), hat(R, x, [1, 0, 0]))
2
```

In other cases, we explicitly want to get the basis vectors and then do something with them. See e.g. `tangent_orthonormal_basis` for `SymmetricPositiveDefinite` (although this currently doesn't seem to work for a random base point and tangent vector). I'm not certain what the use case of this function is.

It would be good to catalog the various uses for ONB's and to figure out a common abstraction. I'm thinking it would be nice for each manifold/metric combo to have a default inexpensive ONB. For methods that use an ONB, one should be able to pass an ONB object (similar to the methods types we've been using).

On the other hand, in some cases it might be useful/efficient to represent a tangent vector as a column vector wrt its ONB, so it might be good to have `TVector` types that explicitly encode this representation.

A few uses for ONBs catalogued here:
- Calculating Jacobians of maps to/from tangent spaces. @mateuszbaran worked out a nice trick for getting the logdet of the Jacobian using SVD, but if we can explicitly create an ONB, that can be much more efficient and AD-friendly. This is necessary for `Retracted` distributions.
- Converting matrix representations of tangent spaces to/from column vector representations. This is important for defining covariance matrices; i.e. a bilinear form on the tangent space that can be interpreted to define a probability measure on the manifold, such as the Riemannian normal.

I'm sure there are many important use cases I'm missing.

**Comment by mateuszbaran** [comment]
> * Calculating Jacobians of maps to/from tangent spaces. @mateuszbaran worked out a nice trick for getting the logdet of the Jacobian using SVD, but if we can explicitly create an ONB, that can be much more efficient and AD-friendly. This is necessary for `Retracted` distributions.

Yes, and to expand on it and for archival purposes, the trick is essentially about constructing ONB using `project_tangent`. Projecting all basis vector of the ambient space gives an overcomplete basis of the tangent space that can be then orthonormalized. Alternatively, one can use SVD or rank-revealing QR (which should be faster but there is apparently no ready implementation in Julia), https://math.berkeley.edu/~mgu/MA273/Strong_RRQR.pdf to rotate the Jacobian to an ONB.

**Comment by kellertuer** [comment]
One thing that I need with the ONB are the eigenvalues of the curvature tensor to compute some of the derivatives that I can use in closed form on some manifolds (in my 1 or 2 implementations the kappa). For that I am much in favour of having such an ONB for as many as possible manifolds (but also the kappas)

**Comment by sethaxen** [comment]
@kellertuer Can you provide a reference for more details on how that type of ONB is derived and how it's used?

**Comment by kellertuer** [comment]
We first used it [here](https://arxiv.org/pdf/1506.02409.pdf) to derive the gradient of a second order difference (Section 3.2). There the ONB (together with the curvature characterising kappa) the ONB is used to evaluate a Jacobi field to get the differential of the geodesic as a function of its start point x (i.e. fixed time t and end point y) – that can be done the same way for
* geodesic and end point
* exp with base and argument (though the second is a function from the tangent space to the manifold)
* log with base and argument 

We tried to explain it also [here](https://arxiv.org/pdf/1807.10090.pdf), Sec 3.1, where the same technique is used to compute the differential of a chain of geodesics (that define the Bezier curve), since we then optimise over the control points of said curve.

Together with the explained adjoint Jacobi field (Sec. 4.2 in the last link for example) one can derive the gradient of a chain of functions involving above functions plus a distance as the last (outer) function.

**Comment by mateuszbaran** [comment]
I think it might be useful to distinguish arbitrary ONBs and ONBs that diagonalize curvature-dependent operators (since they look harder to get). I wonder how much more complicated is obtaining the full Riemann curvature tensor than just that special basis.

**Comment by mateuszbaran** [comment]
I wonder, do we actually need a full ONB (all tangent vectors in the ambient representation that form the basis) for something, or is it just going between vectorized and ambient representation? For the latter, we can just extend the `vee` and `hat` we already have by adding an argument that specifies the type of ONB that is used (for example `ArbitraryONB` when we don't care about special properties and just performance, `DiagonalizingONB` for the thing Ronny uses (which could also return these kappas), `TransportedONB` when we want basis vectors transported along a curve etc.) Is this reasonable?

**Comment by sethaxen** [comment]
> I wonder, do we actually need a full ONB (all tangent vectors in the ambient representation that form the basis) for something, or is it just going between vectorized and ambient representation?

I think for _most_ of my use cases, I indeed only care about interconverting between the vectorized and ambient representations, i.e. something like `vee` and `hat`. For such applications, the basis doesn't even always need to be orthonormal. e.g. `vee` and `hat` for `Rotations` use as the basis the usual generators of the `so(n)` Lie algebra; using our inner product, they're off from an ONB by a factor of √2, but they produce more intuitive vectors (the norm of the vector for `SO(2)`/`SO(3)` is the angle of rotation). So perhaps we want an `AbstractBasis` with subtype `AbstractOrthonormalBasis`.

We could have a `basis(M::Manifold, x, ::AbstractBasis)` method that returns the full ONB.

> For the latter, we can just extend the `vee` and `hat` we already have by adding an argument that specifies the type of ONB that is used

I wonder if more descriptive names than `vee` and `hat` would be useful here, since `vee` and `hat` have specific meanings for kinematic groups but are not necessarily that general. What about `project_basis(M, x, v, AbstractBasis)` and `inverse_project_basis(M, x, v, AbstractBasis)`? 

> for example `ArbitraryONB` when we don't care about special properties and just performance, `DiagonalizingONB` for the thing Ronny uses (which could also return these kappas), `TransportedONB` when we want basis vectors transported along a curve etc

This sounds good to me!

**Comment by mateuszbaran** [comment]
Thanks for the clarification about `vee` and `hat`. You're right that they are not appropriate names here. I'm not sure about the `project_basis` and `inverse_project_basis` either, I wouldn't call it a projection but rather conversion between two representations. Maybe `represent_in_basis` and `inverse_represent_in_basis`? Without actually specifying in the name that we represent tangent vectors since we might want to represent other objects in such bases (for example cotangent vectors). This should probably relate in some way to representations of Christoffel symbols and Riemann tensors.

**Comment by mateuszbaran** [comment]
Implemented in #86 .
